File: clients/node/src/http.ts
// Shared HTTP plumbing: bearer auth, retry with backoff, error mapping.
// Mirrors clients/python/peops_sdk/_http.py. Uses the global `fetch` (Node 18+)
// — zero runtime dependencies.

const RETRYABLE_STATUS = new Set([429, 502, 503, 504]);

export class ApiError extends Error {
  readonly status: number;
  readonly code: string;
  constructor(status: number, code: string, message: string) {
    super(`[${status}] ${code}: ${message}`);
    this.name = "ApiError";
    this.status = status;
    this.code = code;
  }
}

// The hosted PEOps origin every deployment lives behind. Baked in so SDK code
// never has to carry a base URL; override with the PEOPS_BASE_URL env var or an
// explicit baseUrl argument (e.g. for self-host / testing).
export const DEFAULT_BASE_URL = "https://peops.kwon5700.kr";

export function resolveBaseUrl(baseUrl?: string): string {
  return (baseUrl || process.env.PEOPS_BASE_URL || DEFAULT_BASE_URL).replace(/\/+$/, "");
}

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

/** Parse a non-2xx response into an ApiError (reads `detail.code/message`). */
export async function errorFromResponse(resp: Response): Promise<ApiError> {
  let raw = "";
  try {
    raw = await resp.text();
  } catch {
    /* body already consumed / unreadable */
  }
  let code = "error";
  let message = raw.slice(0, 500);
  try {
    const body = JSON.parse(raw) as unknown;
    const detail =
      body && typeof body === "object"
        ? (body as Record<string, unknown>).detail
        : undefined;
    if (detail && typeof detail === "object") {
      const d = detail as Record<string, unknown>;
      code = typeof d.code === "string" ? d.code : "error";
      if (typeof d.message === "string") message = d.message;
    } else if (detail !== undefined) {
      message = String(detail);
    }
  } catch {
    /* not JSON — keep the raw text */
  }
  return new ApiError(resp.status, code, message || resp.statusText);
}

export interface HttpRequestOptions {
  json?: unknown;
  headers?: Record<string, string>;
}

export interface HttpSessionOptions {
  /** Seconds (matches the Python API). */
  timeout?: number;
  maxAttempts?: number;
  maxBackoff?: number;
}

/**
 * fetch wrapper with bearer auth + bounded retry/backoff.
 *
 * Retries transient failures (network errors, 429/5xx) with exponential
 * backoff capped at `maxBackoff`; gives up after `maxAttempts` and raises the
 * last error. 4xx (except 429) never retries.
 */
export class HttpSession {
  readonly baseUrl: string;
  private readonly headers: Record<string, string>;
  private readonly timeoutMs: number;
  private readonly maxAttempts: number;
  private readonly maxBackoff: number;

  constructor(baseUrl: string, apiKey: string, opts: HttpSessionOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.headers = { Authorization: `Bearer ${apiKey}` };
    this.timeoutMs = (opts.timeout ?? 30) * 1000;
    this.maxAttempts = Math.max(1, opts.maxAttempts ?? 3);
    this.maxBackoff = opts.maxBackoff ?? 60;
  }

  async request(
    method: string,
    path: string,
    opts: HttpRequestOptions = {},
  ): Promise<Response> {
    const url = `${this.baseUrl}${path}`;
    const headers: Record<string, string> = { ...this.headers };
    if (opts.json !== undefined) headers["Content-Type"] = "application/json";
    if (opts.headers) Object.assign(headers, opts.headers);

    let lastErr: unknown = null;
    for (let attempt = 0; attempt < this.maxAttempts; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeoutMs);
      let resp: Response | undefined;
      try {
        resp = await fetch(url, {
          method,
          headers,
          body: opts.json !== undefined ? JSON.stringify(opts.json) : undefined,
          signal: controller.signal,
        });
      } catch (err) {
        lastErr = err; // network / abort error — retryable
      } finally {
        clearTimeout(timer);
      }

      if (resp) {
        if (resp.status < 400 || resp.status === 304) return resp;
        if (!RETRYABLE_STATUS.has(resp.status)) {
          throw await errorFromResponse(resp);
        }
        lastErr = await errorFromResponse(resp);
      }

      if (attempt < this.maxAttempts - 1) {
        const backoff = Math.min(this.maxBackoff, 2 ** attempt + Math.random());
        await sleep(backoff * 1000);
      }
    }
    if (lastErr instanceof Error) throw lastErr;
    throw new ApiError(0, "error", String(lastErr));
  }

  // Symmetry with the Python API; there is no persistent client to tear down.
  close(): void {
    /* no-op: global fetch holds no per-session resources */
  }
}

File: clients/python/astra_sdk/_http.py
"""Shared HTTP plumbing: bearer auth, retry with backoff, error mapping."""

from __future__ import annotations

import os
import random
import time
from typing import Any

import httpx

_RETRYABLE_STATUS = {429, 502, 503, 504}

# The hosted Astra origin every deployment lives behind. Baked in so SDK code
# never has to carry a base URL; override with the ASTRA_BASE_URL env var or an
# explicit base_url argument (e.g. for self-host / testing).
DEFAULT_BASE_URL = "https://astra.kwon5700.kr"


def resolve_base_url(base_url: str | None) -> str:
    """The base URL to use: explicit arg → ASTRA_BASE_URL env → hosted default."""
    return (base_url or os.environ.get("ASTRA_BASE_URL") or DEFAULT_BASE_URL).rstrip("/")


class ApiError(Exception):
    """Non-2xx response from the Astra backend."""

    def __init__(self, status: int, code: str, message: str) -> None:
        super().__init__(f"[{status}] {code}: {message}")
        self.status = status
        self.code = code
        self.message = message


def error_from_response(resp: httpx.Response) -> ApiError:
    detail: dict[str, Any] = {}
    try:
        body = resp.json()
        detail = body.get("detail", {}) if isinstance(body, dict) else {}
        if not isinstance(detail, dict):
            detail = {"message": str(detail)}
    except ValueError:
        pass
    return ApiError(
        resp.status_code,
        detail.get("code", "error"),
        detail.get("message", resp.text[:500]),
    )


class HttpSession:
    """httpx.Client wrapper with bearer auth + bounded retry/backoff.

    Retries transient failures (connect errors, 429/5xx) with exponential
    backoff capped at `max_backoff`; gives up after `max_attempts` and raises
    the last error. 4xx (except 429) never retries."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        timeout: float = 30.0,
        max_attempts: int = 3,
        max_backoff: float = 60.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._headers = {"Authorization": f"Bearer {api_key}"}
        self._client = httpx.Client(timeout=timeout)
        self._max_attempts = max(1, max_attempts)
        self._max_backoff = max_backoff

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Any | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        url = f"{self.base_url}{path}"
        merged = dict(self._headers)
        if headers:
            merged.update(headers)
        last_exc: Exception | None = None
        for attempt in range(self._max_attempts):
            try:
                resp = self._client.request(method, url, json=json, headers=merged)
            except httpx.HTTPError as exc:
                last_exc = exc
            else:
                if resp.status_code < 400 or resp.status_code == 304:
                    return resp
                if resp.status_code not in _RETRYABLE_STATUS:
                    raise error_from_response(resp)
                last_exc = error_from_response(resp)
            if attempt < self._max_attempts - 1:
                backoff = min(self._max_backoff, (2 ** attempt) + random.random())
                time.sleep(backoff)
        assert last_exc is not None
        raise last_exc

    def close(self) -> None:
        self._client.close()
